### PR TITLE
 Support GET /token?for=resource for service accounts

### PR DESCRIPTION
**Pull request by alexeykazakov**
- Loads cluster configuration via OpenShift secret
- Uses the cluster auth clients from the cluster configuration to link user accounts
- `/token?for=<cluster_api_url>` returns the corresponding cluster wide token if the request is coming from OSO proxy or Tenant services.

As separate task we will need to add the cluster configuration secret to the OSD side (doesn't block this PR to be merged though).

Fixes https://github.com/fabric8-services/fabric8-auth/issues/179

### CODE REVIEW

**Comment by fabric8cd** [comment]
@alexeykazakov snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-195-1`

**Comment by alexeykazakov** [comment]
[test]

**Comment by alexeykazakov** [comment]
[test]

File: configuration/configuration.go [hunk 1]
```diff
@@ -138,59 +152,78 @@ func NewConfigurationData(mainConfigFile string, serviceAccountConfigFile string
...
+	}
+	c.clusters = map[string]OSOCluster{}
+	for _, cluster := range clusterConf.Clusters {
+		c.clusters[cluster.URL] = cluster
+	}
+
+	return &c, nil
+}
+
+func readFromJSONFile(configFilePath string, defaultConfigFilePath string, configFileName string) (*viper.Viper, error) {
```
Comment (sbryzak) [inline configuration/configuration.go]: Nice refactoring.

File: configuration/configuration.go [hunk 2]
```diff
@@ -838,7 +849,7 @@ OCCAgsB8g8yTB4qntAYyfofEoDiseKrngQT5DSdxd51A/jw7B8WyBK8=
 
 	serviceAccountConfigFileName    = "service-account-secrets.conf"
 	defaultServiceAccountConfigPath = "/etc/fabric8/" + serviceAccountConfigFileName
-)
 
-// ActualToken is actual OAuth access token of github
-var defaultActualToken = strings.Split(camouflagedAccessToken, "-AccessToken-")[0] + strings.Split(camouflagedAccessToken, "-AccessToken-")[1]
+	osoClusterConfigFileName    = "oso-clusters.conf"
+	defaultOsoClusterConfigPath = "/etc/fabric8/" + osoClusterConfigFileName
```
Comment (sbryzak) [inline configuration/configuration.go]: I'm still wondering if we should stick "/etc/fabric8" into a const somewhere

File: openshift/auth.config.yaml [hunk 3]
```diff
@@ -31,6 +29,8 @@ objects:
   data:
     service-account-secrets: >-
       ewogICJhY2NvdW50cyI6IFsKICAgICAgewogICAgICAgICAgIm5hbWUiOiJmYWJyaWM4LXdpdCIsCiAgICAgICAgICAiaWQiOiI1ZGVjNWZkYi0wOWUzLTQ0NTMtYjczZi01YzgyODgzMmIyOGUiLAogICAgICAgICAgInNlY3JldHMiOlsid2l0c2VjcmV0Il0KICAgICAgfSwKICAgICAgewogICAgICAgICAgIm5hbWUiOiJmYWJyaWM4LXRlbmFudCIsCiAgICAgICAgICAiaWQiOiJjMjExZjFiZC0xN2E3LTRmOGMtOWY4MC0wOTE3ZDE2Nzg4OWQiLAogICAgICAgICAgInNlY3JldHMiOlsidGVuYW50c2VjcmV0T2xkIiwgInRlbmFudHNlY3JldE5ldyJdCiAgICAgIH0KICBdCn0=  
+    oso-clusters: >-
+      ewogICJhY2NvdW50cyI6IFsKICAgICAgewogICAgICAgICAgIm5hbWUiOiJmYWJyaWM4LXdpdCIsCiAgICAgICAgICAiaWQiOiI1ZGVjNWZkYi0wOWUzLTQ0NTMtYjczZi01YzgyODgzMmIyOGUiLAogICAgICAgICAgInNlY3JldHMiOlsid2l0c2VjcmV0Il0KICAgICAgfSwKICAgICAgewogICAgICAgICAgIm5hbWUiOiJmYWJyaWM4LXRlbmFudCIsCiAgICAgICAgICAiaWQiOiJjMjExZjFiZC0xN2E3LTRmOGMtOWY4MC0wOTE3ZDE2Nzg4OWQiLAogICAgICAgICAgInNlY3JldHMiOlsidGVuYW50c2VjcmV0T2xkIiwgInRlbmFudHNlY3JldE5ldyJdCiAgICAgIH0KICBdCn0=
```
Comment (sbryzak) [inline openshift/auth.config.yaml]: This is the same value as service-account-secrets.  Perhaps change it to this:

`ew0KICAgICJjbHVzdGVycyI6IFsNCiAgICAgICAgew0KICAgICAgICAgICAgIm5hbWUiOiJ1cy1lYXN0LTIiLA0KICAgICAgICAgICAgInVybCI6Imh0dHBzOi8vYXBpLnN0YXJ0ZXItdXMtZWFzdC0yLm9wZW5zaGlmdC5jb20iLA0KICAgICAgICAgICAgInNlcnZpY2UtYWNjb3VudC10b2tlbiI6ImZYMG5IM2Q2OExRNlNLNXdCRTZRZUtKNlg4QVpHVlFPM2RHUVpaRVRha2htZ21XQXFyMktERlhFNjVLVXdCTzY5YVdvcSIsDQogICAgICAgICAgICAidG9rZW4tcHJvdmlkZXItaWQiOiJmODY3YWMxMC01ZTA1LTQzNTktYTBjNi1iODU1ZWNlNTkwOTAiLA0KICAgICAgICAgICAgImF1dGgtY2xpZW50LWlkIjoiYXV0aGVhc3QyIiwNCiAgICAgICAgICAgICJhdXRoLWNsaWVudC1zZWNyZXQiOiJhdXRoZWFzdDJzZWNyZXQiLA0KICAgICAgICAgICAgImF1dGgtY2xpZW50LWRlZmF1bHQtc2NvcGUiOiJ1c2VyOmZ1bGwiDQogICAgICAgIH0sDQogICAgICAgIHsNCiAgICAgICAgICAgICJuYW1lIjoidXMtZWFzdC0yYSIsDQogICAgICAgICAgICAidXJsIjoiaHR0cHM6Ly9hcGkuc3RhcnRlci11cy1lYXN0LTJhLm9wZW5zaGlmdC5jb20iLA0KICAgICAgICAgICAgInNlcnZpY2UtYWNjb3VudC10b2tlbiI6ImFrNjFUNlJTQWFjV0ZydWgxdlpQOGN5VU9CdFEzQ2h2MXJkT0JkZFN1YzluWjJ3RWNzODFESFhSTzU1TnBJcFZROHVpSCIsDQogICAgICAgICAgICAidG9rZW4tcHJvdmlkZXItaWQiOiI4ODZjN2VhMy1lZjk3LTQ0M2QtYjM0NS1kZTk0Yjk0YmI2NWQiLA0KICAgICAgICAgICAgImF1dGgtY2xpZW50LWlkIjoiYXV0aGVhc3QyYSIsDQogICAgICAgICAgICAiYXV0aC1jbGllbnQtc2VjcmV0IjoiYXV0aGVhc3QyYXNlY3JldCIsDQogICAgICAgICAgICAiYXV0aC1jbGllbnQtZGVmYXVsdC1zY29wZSI6InVzZXI6ZnVsbCINCiAgICAgICAgfQ0KICAgIF0NCn0=`

File: openshift/auth.app.yaml [hunk 4]
```diff
@@ -162,13 +152,22 @@ objects:
...
         volumes:
         - name: auth-serviceaccounts
           secret:
             secretName: fabric8-auth-service-accounts
             items:
             - key: service-account-secrets
               path: service-account-secrets.conf
+        - name: auth-osoclusters
+          secret:
+            secretName: fabric8-auth-oso-clusters
```
Comment (sbryzak) [inline openshift/auth.app.yaml]: This is not consistent with the secrets configuration.  It should still be fabric8-auth-service-accounts, unless you create a new, separate configuration for fabric8-auth-oso-clusters.

File: openshift/auth.config.yaml [hunk 5]
```diff
@@ -31,6 +29,8 @@ objects:
   data:
     service-account-secrets: >-
       ewogICJhY2NvdW50cyI6IFsKICAgICAgewogICAgICAgICAgIm5hbWUiOiJmYWJyaWM4LXdpdCIsCiAgICAgICAgICAiaWQiOiI1ZGVjNWZkYi0wOWUzLTQ0NTMtYjczZi01YzgyODgzMmIyOGUiLAogICAgICAgICAgInNlY3JldHMiOlsid2l0c2VjcmV0Il0KICAgICAgfSwKICAgICAgewogICAgICAgICAgIm5hbWUiOiJmYWJyaWM4LXRlbmFudCIsCiAgICAgICAgICAiaWQiOiJjMjExZjFiZC0xN2E3LTRmOGMtOWY4MC0wOTE3ZDE2Nzg4OWQiLAogICAgICAgICAgInNlY3JldHMiOlsidGVuYW50c2VjcmV0T2xkIiwgInRlbmFudHNlY3JldE5ldyJdCiAgICAgIH0KICBdCn0=  
+    oso-clusters: >-
```
Comment (sbryzak) [inline openshift/auth.config.yaml]: As per my previous comment, this secret value is defined under the fabric8-auth-service-accounts secret.  I would be inclined to put this value in a separate secret configuration, like so:

```
- apiVersion: v1
  kind: Secret
  metadata:
    name: fabric8-auth-oso-clusters
  type: Opaque
  data:
    oso-clusters: >-
      ewogICJhY2NvdW50cyI6IFsKICAgICAgewogICAgICAgICAgIm5hbWUiOiJmYWJyaWM4LXdpdCIsCiAgICAgICAgICAiaWQiOiI1ZGVjNWZkYi0wOWUzLTQ0NTMtYjczZi01YzgyODgzMmIyOGUiLAogICAgICAgICAgInNlY3JldHMiOlsid2l0c2VjcmV0Il0KICAgICAgfSwKICAgICAgewogICAgICAgICAgIm5hbWUiOiJmYWJyaWM4LXRlbmFudCIsCiAgICAgICAgICAiaWQiOiJjMjExZjFiZC0xN2E3LTRmOGMtOWY4MC0wOTE3ZDE2Nzg4OWQiLAogICAgICAgICAgInNlY3JldHMiOlsidGVuYW50c2VjcmV0T2xkIiwgInRlbmFudHNlY3JldE5ldyJdCiAgICAgIH0KICBdCn0=

```

Comment (alexeykazakov) [reply openshift/auth.config.yaml]: Good catch. Maybe we should move all the secret values to the same “auth” secret instead? WDYT? Including all values that we mount as env vars and config files. Or alternatively one secret for env var values and one (auth-config-files?) for service accounts and oso clusters.
WDYT?

**Comment by fabric8cd** [comment]
@alexeykazakov snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-195-2`

**Comment by fabric8cd** [comment]
@alexeykazakov snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-195-3`

File: openshift/auth.config.yaml [hunk 6]
```diff
@@ -20,9 +20,9 @@ objects:
     github.client.id: Cg==
     github.client.secret: Cg==
     oso.client.apiurl: aHR0cHM6Ly9hcGkuY29uc29sZS5zdGFydGVyLXVzLWVhc3QtMi5vcGVuc2hpZnQuY29t
-    oso.client.id: Cg==
-    oso.client.secret: Cg==
     tenantsvc.url: aHR0cDovL2Y4dGVuYW50Ojgw
+    oso-clusters: >-
```
Comment (vpavlin) [inline openshift/auth.config.yaml]: Can you please change this to `oso.cluster` to keep the naming of other keys?

File: openshift/auth.app.yaml [hunk 7]
```diff
@@ -162,13 +152,22 @@ objects:
...
           secret:
             secretName: fabric8-auth-service-accounts
             items:
             - key: service-account-secrets
               path: service-account-secrets.conf
+        - name: auth-osoclusters
+          secret:
+            secretName: auth
+            items:
+            - key: oso-clusters
```
Comment (vpavlin) [inline openshift/auth.app.yaml]: And do not forget to change it here as well

File: openshift/auth.config.yaml (hunk 6 above)
Comment (alexeykazakov) [reply openshift/auth.config.yaml]: Good point! Created a new PR for that - https://github.com/fabric8-services/fabric8-auth/pull/197

**Pull request merged**